src/main/java/com/google/devtools/build/remote/client/AbstractRemoteActionCache.java

```diff
@@ -15,17 +15,184 @@
 package com.google.devtools.build.remote.client;
 
 import com.google.devtools.remoteexecution.v1test.Digest;
+import com.google.devtools.remoteexecution.v1test.Directory;
+import com.google.devtools.remoteexecution.v1test.DirectoryNode;
+import com.google.devtools.remoteexecution.v1test.FileNode;
+import com.google.devtools.remoteexecution.v1test.Tree;
+import com.google.protobuf.ByteString;
 import java.io.IOException;
 import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
 
 /** A cache for storing artifacts (input and output) as well as the output of running an action. */
 public abstract class AbstractRemoteActionCache {
+
+  public DigestUtil getDigestUtil() {
+    return digestUtil;
+  }
+
   protected final DigestUtil digestUtil;
 
   public AbstractRemoteActionCache(DigestUtil digestUtil) {
     this.digestUtil = digestUtil;
   }
 
-  /** Downloads a single blob with the specified digest into an output stream. */
+  /**
+   * Download a tree with the {@link Directory} given by digest as the root directory of the tree.
+   *
+   * @param rootDigest The digest of the root {@link Directory} of the tree
+   * @return A tree with the given directory as the root.
+   * @throws IOException in the case that retrieving the blobs required to reconstruct the tree
+   *     failed.
+   */
+  public Tree getTree(Digest rootDigest) throws IOException {
+    byte directoryBytes[];
```

Comment:
Nit: declare a variable right before its use. Oftentimes, this will result in not even needing the variable.
This is better:

byte directoryBytes = downloadBlob(rootDigest);
Directory dir = Directory.parseFrom(directoryBytes);
return getTree(dir);

But you can also make it a one-liner:
return getTree(Directory.parseFrom(downloadBlob(rootDigest)));

Reply:
Done